Use date-fns and date-fns-tz instead of moment

vcanales opened the pull request:
## Description
This PR refactors @wordpress/date to use `date-fns` instead of momentjs for date formatting. `date-fns` has a few advantages over moments js, including being half the size, and also using Browser APIs to handle timezones through `date-fns-tz`—which can be polyfilled if necessary.

## How has this been tested?

Having the existing tests pass.

### Make the PHP > Moment format map work with date-fns formats instead.

- [x] Day format tokens.
- [x] Week format tokens.
- [x] Month format tokens.
- [x] Format function tokens.
- [x] Localization.

### Refactored refactored functions.
I've been refactoring the internal dependencies of these functions as a part of this.
- [x] `format`
- [x] `date`
- [x] `dateI18n`
- [x] `gmdate` 
- [x] `gmdateI18n`
- [x] `isInTheFuture`
- [x] `getDate`

### Missing things
- [x] Transforming integer offsets into valid timezone offsets ie. turning `-4` into `-0400` or `-04:00`


## Screenshots <!-- if applicable -->

## Types of changes
Refactoring date functions, having the tests pasts as they exist right now.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

swissspidy commented:
Does localization still work as expected?

See https://github.com/WordPress/wordpress-develop/blob/de73a6a8e9a6e2546e0d3624c2cb0daf4d96a819/src/wp-includes/script-loader.php#L326-L364 and https://github.com/WordPress/wordpress-develop/blob/de73a6a8e9a6e2546e0d3624c2cb0daf4d96a819/src/wp-includes/script-loader.php#L140-L167 for example.

vcanales commented:
> Does localization still work as expected?

I haven't taken a look yet, but [`date-fns`'s localization works differently](https://date-fns.org/v2.16.1/docs/Locale), so the JSON object being created there will have to be changed. Thanks for pointing me there!

swissspidy commented:
If it helps, I did the same exercise for the date-fns locale stuff in https://github.com/google/web-stories-wp/tree/main/assets/src/date. Might be helpful here :)

vcanales commented:
Got all unit tests passing, locally at least. There's still a dependency problem, particularly with `date-fns-tz` that is still making things fail up here—not sure how to fix that.

### Todo
- [ ] Get custom/function format tokens working. Some of these were created specifically because moment didn't support those formats.
- [ ] Reduce repetition: I'm calling `date-fns`'s  `format` function directly passing the whole options object each time, but there's things that can be moved to a main wrapper function, which should replace the current `format` function which was built to work with `momentjs`.

vcanales commented:
Tests are passing now, but there are still some things missing, so I'll add a few more to cover those missing cases. ie. tests shouldn't really be passing right now, because there's a whole set of format tokens that are not implemented.

- [ ] Write missing test cases.

Commit:
Implement zero-indexed day of the week

Commit:
Implement epoc format

Commit:
Add missing timezone parameter

Commit:
Add missing jsdoc for internal functions

vcanales commented:
Marking this as ready for review. I'd appreciate some advice on further testing this code if it is necessary, as tests are passing but I'm not a 100% sure of how thorough they are.

gziolo commented:
You can test it by playing with the publish date of the post that uses the calendar. Latest Posts block is another one that should use dates.

Commit:
Use zonedTimeToUtc on post scheduler label

The post's date attribute is a datetime string with no indication of which timezone
it is in, so in order for it to be correctly read by the formatting functions it has to
be transformed to its UTC equivalent relative to the site's timezone.

jorgefilipecosta reviewed (commented):
Formats 'c' and 'r', crash the editor e.g: calling wp.date.dateI18n('c', '2020-10-16T14:36:58'); on the console crashes.
This crash is noticeable for example if we add the latest posts block and then enable to show the post dates.
I guess we need some updates on string 'yyyy-MM-DDTHH:mm:ssZ' e.g: DD should be d. Z should be xxx.
I also notice some formats don't behave as expected (equivalent to the PHP version). The output of wp.date.format('MM', '2020-10-16T14:36:58'); before was OutOut now is October.
If I do  wp.date.format('FF', '2020-10-16T14:36:58'); the output before was OctoberOctober now it is only October.

vcanales commented:
@jorgefilipecosta Thank you for the review, I'll look into these comments!

jorgefilipecosta commented on packages/date/src/index.js:
```diff
@@ -470,58 +610,8 @@ export function isInTheFuture( dateValue ) {
 }
 
-setupWPTimezone();
+export { zonedTimeToUtc };
```
What makes us need to export zonedTimeToUtc?

vcanales replied:
Only the one call that you mentioned bellow; I will remove this export once we decide which way to go with using timezones for I18n dates.

jorgefilipecosta commented on packages/editor/src/components/post-schedule/label.js:
```diff
@@ -2,16 +2,23 @@
+		? dateI18n(
 				`${ settings.formats.date } ${ settings.formats.time }`,
-				date
+				zonedTimeToUtc( date, getActualTimezone() )
```
Simply formatting a date that comes in the WordPress timezone, and keeping in the WordPress seems like a very common. Developers should not need to call zonedTimeToUtc and getActualTimezone. It seems we don't have a function that formats a date that comes in the WordPress timezone (inexplicitly like the dates from WordPress API) and formats it keeping the WordPress timezone. On https://github.com/WordPress/gutenberg/pull/26276 I'm proposing a new function that does that (and should probably the function used in 99% of the cases where date formatting in WordPress is needed). What do you think about the new function, do you think we need something new or we can adapt what we have here?

vcanales replied:
I think we should be able to move this operation (`zonedTimeToUtc( date, getActualTimezone() )`) into the `dateI18n` function under a parameter, or even make it the default way in which it runs, which would make `dateI18n` work as the one proposed on #26276. Either way, I think that using the current `timezone` parameter on `dateI18n`, and looking for timezone fragments of the timestamp can help us get there.

>  Developers should not need to call zonedTimeToUtc and getActualTimezone.

Agreed! We could either add the function proposed or do what I mention above instead of having developers call these methods manually.

vcanales commented:
> I also notice some formats don't behave as expected (equivalent to the PHP version). The output of wp.date.format('MM', '2020-10-16T14:36:58'); before was OutOut now is October.
> If I do wp.date.format('FF', '2020-10-16T14:36:58'); the output before was OctoberOctober now it is only October.

Regarding this, the fix I came up with is a bit convoluted, but the justification for it is explained in the [commit message](https://github.com/WordPress/gutenberg/pull/25782/commits/23c32191cbccb2d6da07e6319f9cbe6c579f954b):

> PHP Format tokens don't 'compose', meaning that multiple identical
tokens in a row just concatenate the result instead of generating a new
format.

> With momentjs it was possible to separate identical tokens by escaping an empty
character, ie. "[]" would add invisible space between the two formatting tokens;
with date-fns  that is not possible, as two quote in a row (the escaping character in date-fns) equals an actual quote.

I landed on adding a unicode character to separate formatting tokens, and then replacing it out of the resulting format. I can't really see an alternative, easy way out of this limitation. I think that a unicode character such as the one I chose, which is a [six-per-em space](https://en.wikipedia.org/wiki/Whitespace_character#Unicode), is obscure enough to not collide with any actual usage of the formatting functions. This can be changed for any other arbitrary string as well.

Aside from those considerations, the 3 issues you mentioned with formatting tokens are fixed. I also added tests for them to make sure of that.

I'd love to hear your thoughts on these changes, @jorgefilipecosta.

swissspidy commented on packages/date/src/index.js:
```diff
@@ -226,50 +201,54 @@ const formatMap = {
+	L( dateValue ) {
+		return isLeapYear( dateValue ) ? '1' : '0';
 	},
 	o: 'GGGG',
```
Shouldn't this be `RRRR`?

jorgefilipecosta replied:
Yes, it seems it should be RRRR according to the docs of the last version https://date-fns.org/v2.16.1/docs/format. But on previous versions, it was GGGG https://date-fns.org/v1.30.1/docs/format. This type of breaking change is dangerous we should be very careful about each package version upgrade.

vcanales replied:
As a part of applying these formatting token solutions, I'll write tests for all available formats; this way we'll be able to catch these breaking errors if they happen when updating `date-fns`.

swissspidy commented on packages/date/src/index.js:
```diff
@@ -314,58 +315,197 @@ const formatMap = {
+		} );
+	},
+	U( dateValue ) {
+		return getUnixTime( utcToZonedTime( dateValue, getActualTimezone() ) );
```
Could just use `t`?

jorgefilipecosta replied:
According to the docs, it seems the 't' format would work well https://date-fns.org/v1.30.1/docs/format.

swissspidy commented on packages/date/src/index.js:
```diff
@@ -281,30 +260,52 @@ const formatMap = {
+	Z( dateValue ) {
 		// Timezone offset in seconds.
-		const offset = momentDate.format( 'Z' );
+		const offset = dateFnsFormat(
+			utcToZonedTime( dateValue, 'UTC' ),
+			'XXX'
+		);
```
Since this is going to be something like `+02:00` you want to cast `parts` to `Number`s.

Otherwise right now `parts` will be something like `[ 02, 00 ]`, which are interpreted as octals, making the calculation below incorrect.

swissspidy replied:
```js
const parts = offset.substring(1).split(':').map(Number);
```

should work

vcanales replied:
Thanks! I've applied your suggestion.

jorgefilipecosta commented:
I'm getting some errors while trying to execute this branch. Even after npm install.


Am I missing something?
```
[0] ERROR in ./node_modules/date-fns-tz/esm/zonedTimeToUtc/index.js
[0] Module not found: Error: Can't resolve 'date-fns/esm/_lib/cloneObject' in '/Users/pc/dev/core/wordpress-develop/build/wp-content/plugins/gutenberg/node_modules/date-fns-tz/esm/zonedTimeToUtc'
[0]  @ ./node_modules/date-fns-tz/esm/zonedTimeToUtc/index.js 1:0-55 33:24-35
[0]  @ ./node_modules/date-fns-tz/esm/index.js
[0]  @ ./packages/date/build-module/index.js
[0] 
[0] ERROR in ./node_modules/date-fns-tz/esm/toDate/in
[0] dex.js
[0] Module not found: Error: Can't resolve 'date-fns/esm/_lib/getTimezoneOffsetInMilliseconds/index.js' in '/Users/pc/dev/core/wordpress-develop/build/wp-content/plugins/gutenberg/node_modules/date-fns-tz/esm/toDate'
[0]  @ ./node_modules/date-fns-tz/esm/toDate/index.js 2:0-104 173:15-46 174:15-46
[0]  @ ./node_modules/date-fns-tz/esm/index.js
[0]  @ ./packages/date/build-module/index.js
[0] 
[0] ERROR in ./node_modules/date-fns-tz/esm/toDate/index.js
[0] Module not found: Error: Can't resolve 'date-fns/esm/_lib/toInteger/index.js' in '/Users/pc/dev/core/wordpress-develop/build/wp-content/plugins/gutenberg/node_modules/date-fns-tz/esm/toDate'
[0]  @ ./node_modules/date-fns-tz/esm/toDate/index.js 1:0-60 100:8-17
[0]  @ ./node_modules/date-fns-tz/esm/index.js
[0]  @ ./packages/date/build-module/index.js
[0] 
[0] ERROR in ./node_modules/date-fns-tz/esm/format/index.js
[0] Module not found: Error: Can't resolve 'date-fns/esm/format' in '/Users/pc/dev/core/wordpress-develop/build/wp-content/plugins/gutenberg/node_modules/date-fns-tz/esm/format'
[0]  @ ./node_modules/date-fns-tz/esm/format/index.js 1:0-47 334:9-22
[0]  @ ./node_modules/date-fns-tz/esm/index.js
[0]  @ ./packages/date/build-module/index.js
[0] 
[0] ERROR in ./node_modules/date-fns-tz/esm/zonedTimeToUtc/index.js
[0] Module not found: Error: Can't resolve 'date-fns/esm/format' in '/Users/pc/dev/core/wordpress-develop/build/wp-content/plugins/gutenberg/node_modules/date-fns-tz/esm/zonedTimeToUtc'
[0]  @ ./node_modules/date-fns-tz/esm/zonedTimeToUtc/index.js 2:0-40 31:11-17
[0]  @ ./node_modules/date-fns-tz/esm/index.js
[0]  @ ./packages/date/build-module/index.js
[0] 
[0] ERROR in ./node_modules/date-fns-tz/esm/utcToZonedTime/index.js
[0] Module not found: Error: Can't resolve 'date-fns/esm/subMilliseconds' in '/Users/pc/dev/core/wordpress-develop/build/wp-content/plugins/gutenberg/node_modules/date-fns-tz/esm/utcToZonedTime'
[0]  @ ./node_modules/date-fns-tz/esm/utcToZonedTime/index.js 2:0-58 46:6-21
[0]  @ ./node_modules/date-fns-tz/esm/index.js
[0]  @ ./packages/date/build-module/index.js
```

Commit:
Export getActualTimezone

vcanales commented:
> I'm getting some errors while trying to execute this branch. Even after npm install.
> 
> Am I missing something?

No, not missing anything at all; I'm working on getting this branch running again, but having some probably dependency relate issues at the moment. Apologies for the inconvenience, I'll let you know when we can test again.

Commit:
Use date-fns-tz

Commit:
Add @todo comments

Commit:
Reafactor all non-localized formatting functions

Commit:
Get all tests passing

Commit:
Add missing type on isInTheFuture fn

Commit:
Parse date string before use

Commit:
Create date object instead of trying to parse ISO

Commit:
Remove escaping characters from format translation

Commit:
Expect uppercase meridian indicator

Commit:
Implement 5 out of 7 custom formattings tokens

Commit:
Remove unnecessary timezone operations

Commit:
Implement timezone indicator formatting tokens

Commit:
Update jsdocs

Commit:
Use 't' to get epoch format

Commit:
Move timezone offset fix  to the datei18n function

Commit:
Fix reset button timezone

Commit:
Refactor floating post date logic

No need to factor in null, as dateI18n defaults to `now`.

Commit:
Fix empty time inputs on reset

Commit:
Add tests for most formatting tokens

Skipping the ones who, as a result of adding tests,
seem to be misbehaving.

Commit:
Go back to lower case meridiem indicator

mkaz reviewed (approved):
Ok, looks like the reset is fixed and it is all behaving properly! 👍 

I tested switching the site to UTC and other timezones different than my own and it shows properly in the selector what timezone I'm setting the post as.

When setting the timezone to the same as my browser no timezone hint shows as it should be.

Thanks for sticking with it all, I know how much of a headache time zones are

vcanales commented:
Unrelated e2e test failure aside, this is ready to merge.

Tracking subsequent tasks on the following issues:
- https://github.com/WordPress/gutenberg/issues/26862
- https://github.com/WordPress/gutenberg/issues/26836

gziolo commented:
First of all, awesome work finishing this PR. I appreciate all the efforts to make it happen 🎉 

I'm experiencing some test failures in my local env when using the latest `master`:

```
Summary of all failing tests
 FAIL  packages/date/src/test/index.js
  ● PHP Format Tokens › should support "z" to obtain zero-indexed day of the year

    expect(received).toBe(expected) // Object.is equality

    Expected: "0"
    Received: "364"

      240 |             const formattedDate = dateNoI18n( 'z', '2019-01-01' );
      241 | 
    > 242 |             expect( formattedDate ).toBe( '0' );
          |                                     ^
      243 |     } );
      244 | 
      245 |     it( 'should support "W" to obtain ISO-8601 week number of year, weeks starting on Monday', () => {

      at Object.<anonymous> (packages/date/src/test/index.js:242:27)

  ● PHP Format Tokens › should support "t" to obtain the days in a given month

    expect(received).toBe(expected) // Object.is equality

    Expected: "28"
    Received: "31"

      276 |             const formattedDate = dateNoI18n( 't', '2019-02' );
      277 | 
    > 278 |             expect( formattedDate ).toBe( '28' );
          |                                     ^
      279 |     } );
      280 | 
      281 |     it( 'should support "L" to obtain whether or not the year is a leap year', () => {

      at Object.<anonymous> (packages/date/src/test/index.js:278:27)

  ● PHP Format Tokens › should support "L" to obtain whether or not the year is a leap year

    expect(received).toBe(expected) // Object.is equality

    Expected: "1"
    Received: "0"

      282 |             const formattedDate = dateNoI18n( 'L', '2020' );
      283 | 
    > 284 |             expect( formattedDate ).toBe( '1' );
          |                                     ^
      285 |     } );
      286 | 
      287 |     it( 'should support "o" to obtain the ISO-8601 week-numbering year. This has the same value as Y, except that if the ISO week number (W) belongs to the previous or next year, that year is used instead.', () => {

      at Object.<anonymous> (packages/date/src/test/index.js:284:27)


Test Suites: 1 failed, 429 passed, 430 total
Tests:       3 failed, 2 skipped, 4628 passed, 4633 total
```

It surely is related to a difference between timezones on machines, because it passes on CI.

vcanales commented:
Thanks for bringing this up; although I'm not 100% if it'll be useful, could you share your timezone with me? I tested this with different times locally before merging. Any other details about your setup would be appreciated as well.

gziolo commented:
`Time Zone: Europe/Warsaw` - if that helps. Does the test suite set a particular timezone maybe to avoid those type of issues?

gziolo commented on packages/date/src/index.js:
```diff
@@ -75,56 +85,20 @@ let settings = {
  * @param {Object} dateSettings Settings, including locale data.
  */
 export function setSettings( dateSettings ) {
 	settings = dateSettings;
```
It might be necessary to set all those settings directly to `moment` script handle that some plugins still use. See https://github.com/WordPress/gutenberg/issues/27159 that reports error on `tz` property.

gziolo replied:
I have more details included in https://github.com/WordPress/gutenberg/issues/27159#issuecomment-731651670.

vcanales replied:
> I have more details included in [#27159 (comment)](https://github.com/WordPress/gutenberg/issues/27159#issuecomment-731651670).

Thanks for taking a look, I'll do the same and reply on the issue you linked.

jorgefilipecosta commented:
Hi @vcanales,
I'm noticing a regression. The timezone of the test website is UTC-5 more especially Detroit, my local timezone is Lisbon on current DST equal to UTC.
When I try to schedule a post to be published in one hour after the current website date the publish button says publish instead of schedule the same if I publish five hours after. Only if I publish six hours after it says schedule.

vcanales commented:
@jorgefilipecosta Thanks for bringing it up; I'll take a look.

jorgefilipecosta commented:
Hi @vcanales, I'm trying to revive and update https://github.com/WordPress/gutenberg/pull/26276 (to make it take into account the moment removal). I think it will fix this issue :)

vcanales commented:
> Hi @vcanales, I'm trying to revive and update #26276 (to make it take into account the moment removal). I think it will fix this issue :)

Great! I'll keep a an eye on that PR. Let me know if I can help too!

jorgefilipecosta commented on packages/date/src/index.js:
```diff
@@ -426,9 +549,22 @@ export function dateI18n( dateFormat, dateValue = new Date(), timezone ) {
-	dateMoment.locale( settings.l10n.locale );
-	return format( dateFormat, dateMoment );
+	return formatTZ(
+		utcToZonedTime(
```
Hi @vcanales, We are passing a manipulated date value to formatTZ, and the original value to translateFormat. translateFormat for some formats like 'c' outputs the date directly. I think the date value we pass to formatTZ and to translateFormat should always be the same, otherwise depending on if it is a mapped format or a format that is a function we output different things. In PR https://github.com/WordPress/gutenberg/pull/27272/ I added unit tests that show this bug happening.

jorgefilipecosta commented:
Hi @vcanales I found some issues in the functions we already have (dateI18n, format etc) while trying to implement dateI18nAddTimezone. I think the root cause is translateFormat. I created some unit tests that show these bugs happening https://github.com/WordPress/gutenberg/pull/27272.
Let me know if you think the output I'm expecting on the unit tests is wrong. I guess we should try to fix these issues on the current functions before adding the new one.

vcanales commented:
Hey @jorgefilipecosta. I think at the root of the issues is that since `date-fns` works with a functional approach, by using it as we are right now we're missing on the benefit of having to setup the timezone in a single place. When developing this I had (tried?) to compensate by embedding timezone logic into the format translation. I'll revisit this based on your tests, as the output seems correct.

vcanales commented:
> Let me know if you think the output I'm expecting on the unit tests is wrong. I guess we should try to fix these issues on the current functions before adding the new one.

I also want to add that, while the output is correct, we might want to change the timestamps we test against to not include any timezone suffixes, since the API does not provide those. I added a more "in depth" explanation of this on https://github.com/WordPress/gutenberg/pull/27002#discussion_r530364520
